src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/Grpc/GrpcStatusDeserializer.cs

```diff
@@ -0,0 +1,312 @@
+{
+#pragma warning disable SA1310 // Field names should not contain underscore
+    // Wire types in protocol buffers
+    private const int WIRETYPE_VARINT = 0;
+    private const int WIRETYPE_FIXED64 = 1;
+    private const int WIRETYPE_LENGTH_DELIMITED = 2;
+    private const int WIRETYPE_FIXED32 = 5;
+#pragma warning restore SA1310 // Field names should not contain underscore
+
+    internal static TimeSpan? TryGetGrpcRetryDelay(string? grpcStatusDetailsHeader)
+    {
+        try
+        {
+            var retryInfo = ExtractRetryInfo(grpcStatusDetailsHeader);
+            if (retryInfo?.RetryDelay != null)
+            {
+                return TimeSpan.FromSeconds(retryInfo.Value.RetryDelay.Value.Seconds) +
+                       TimeSpan.FromTicks(retryInfo.Value.RetryDelay.Value.Nanos / 100); // Convert nanos to ticks
+            }
+        }
+        catch
+        {
+            // TODO: Log exception to event source.
+            return null;
+        }
+
+        return null;
+    }
+
+    // Marked as internal for test.
+    internal static Status? DeserializeStatus(string? grpcStatusDetailsBin)
+    {
+        if (string.IsNullOrWhiteSpace(grpcStatusDetailsBin))
+        {
+            return null;
+        }
+
+        var status = new Status();
+        byte[] data = Convert.FromBase64String(grpcStatusDetailsBin);
+        using (var stream = new MemoryStream(data))
+        {
+            while (stream.Position < stream.Length)
+            {
+                var tag = DecodeTag(stream);
+                var fieldNumber = tag >> 3;
+                var wireType = tag & 0x7;
+
+                switch (fieldNumber)
+                {
+                    case 1: // code
+                        status.Code = DecodeInt32(stream);
+                        break;
+                    case 2: // message
+                        status.Message = DecodeString(stream);
+                        break;
+                    case 3: // details
+                        status.Details.Add(DecodeAny(stream));
+                        break;
+                    default:
+                        SkipField(stream, wireType);
+                        break;
+                }
+            }
+        }
+
+        return status;
+    }
+
+    // Marked as internal for test.
+    internal static RetryInfo? ExtractRetryInfo(string? grpcStatusDetailsBin)
+    {
+        var status = DeserializeStatus(grpcStatusDetailsBin);
+        if (status == null)
+        {
+            return null;
+        }
+
+        foreach (var detail in status.Value.Details)
+        {
+            if (detail.TypeUrl != null && detail.TypeUrl.EndsWith("/google.rpc.RetryInfo"))
+            {
+                return DeserializeRetryInfo(detail.Value!);
+            }
+        }
+
+        return null;
+    }
+
+    private static RetryInfo? DeserializeRetryInfo(byte[] data)
+    {
+        RetryInfo? retryInfo = null;
+        using (var stream = new MemoryStream(data))
```

Comment:
I see a lot of opportunity in this code to make it more efficient. For example, why allocate a `MemoryStream` and not operate on a `Span<byte>`? But even with that, where does this `byte[]` come from? HttpClient response content? That is probably already a stream which could be passed in to this code. Probably don't need to force it to be a `byte[]` at all 😄

But I'm not going to block this. We can improve it later. Guessing this response stuff is only executed once per transmission so some small number of allocations won't be the end of the world.

Reply:
It's a good call out and we could improve later as follow up.

> where does this byte[] come from? HttpClient response content?
We read string from http header and need to convert it to byte.